experiments/scan_classes/sequences.py

```diff
@@ -0,0 +1,261 @@
+                        if b_pulse['combined_pulse_index'] == pulse['combined_pulse_index']:
+                            combined_pulse_2.append(b_pulse)
+
+                    for pulse_1, pulse_2 in zip(combined_pulse_1, combined_pulse_2):
+                        # combine_index should be the same because they were in order.
+                        if pulse_1['slice_id'] != pulse_2['slice_id']:
+                            pulse['isarepeat'] = False
+                            break
+                        if pulse_1['intra_pulse_start_time'] != pulse_2['intra_pulse_start_time']:
+                            pulse['isarepeat'] = False
+                            break
+                        if pulse_1['pulse_shift'] != pulse_2['pulse_shift']:
+                            pulse['isarepeat'] = False
+                            break
+                    else:  # no break
+                        pulse['isarepeat'] = True
+
+        last_pulse = self.pulses[-1]
+        this_pulse_len = 0
+        if last_pulse['iscombined']:
+            for pulse_ind in range(-(last_pulse['combine_total']), 0):
+                this_pulse = self.pulses[pulse_ind]
+                if (this_pulse['pulse_len'] + this_pulse['intra_pulse_start_time']) > this_pulse_len:
+                    this_pulse_len = this_pulse['pulse_len'] + this_pulse['intra_pulse_start_time']
+        else:
+            this_pulse_len = last_pulse['pulse_len']
+
+        self.last_pulse_len = this_pulse_len
+
+
+        # FIND the max scope sync time
+        # 19 is the sample delay below; how do we calculate this? # REVIEW #6 TODO find out the reason for these magic numbers
+        self.ssdelay = 0  # ssdelay is time required to wait after last pulse.
+        for slice_id in self.slice_ids:
+            newdelay = (self.slice_dict[slice_id]['nrang'] + 19 + 10) * self.slice_dict[slice_id]['pulse_len']
+            if newdelay > self.ssdelay:
+                self.ssdelay = newdelay
+        # The delay is long enough for any slice's pulse length and nrang to be accounted for.
+
+        # FIND the sequence time. Time before the first pulse is 70 us when RX and TR set up for the first pulse. The
+        # timing to the last pulse is added, as well as its pulse length and the RX/TR delay at the end of last pulse.
+        self.seqtime = self.options.atten_window_time_start + 2*self.options.tr_window_time + \
+                       self.options.atten_window_time_end + self.pulses[-1]['pulse_timing_us'] + \
+                       self.last_pulse_len
+
+        # FIND the total scope sync time and number of samples to receive.
+        self.sstime = self.seqtime + self.ssdelay
+        self.numberofreceivesamples = int(self.options.rx_sample_rate * self.sstime * 1e-6)
+
+    def build_pulse_transmit_data(self, slice_to_beamdir_dict, txctrfreq, txrate, options):  #TODO only take in things you need or add needed params from options in the init function.
```

Comment:
Ya, I have some ideas for making borealis more portable that we could talk about. I was thinking instead of putting main_antenna_count and main_antenna_spacing in config, put a transmit antenna mapping with some x and y orientation data (ie not assuming all transmit antennas are in a row at a certain spacing). These functions would then need to change and we would have to pass orientation data to signal processing block as well